glean-core/ffi/src/byte_buffer.rs

```diff
@@ -0,0 +1,152 @@
+
+//! ByteBuffer is a struct that represents an array of bytes to be sent over the FFI boundaries.
+//!
+//! This is a copy of the same struct from [ffi-support],
+//! with the difference that the length is encoded as a `i32`.
+//!
+//! [ffi-support]: https://docs.rs/ffi-support/0.4.0/src/ffi_support/lib.rs.html#390-393
+
+/// ByteBuffer is a struct that represents an array of bytes to be sent over the FFI boundaries.
+/// There are several cases when you might want to use this, but the primary one for us
+/// is for returning protobuf-encoded data to Swift and Java. The type is currently rather
+/// limited (implementing almost no functionality), however in the future it may be
+/// more expanded.
+///
+/// ## Caveats
+///
+/// Note that the order of the fields is `len` (an i32) then `data` (a `*mut u8`), getting
+/// this wrong on the other side of the FFI will cause memory corruption and crashes.
+/// `i32` is used for the length instead of `u64` and `usize` because JNA has interop
+/// issues with both these types.
+///
+/// ByteBuffer does not implement Drop. This is intentional. Memory passed into it will
+/// be leaked if it is not explicitly destroyed by calling [`ByteBuffer::destroy`]. This
+/// is because in the future, we may allow it's use for passing data into Rust code.
+/// ByteBuffer assuming ownership of the data would make this a problem.
+///
+/// Note that calling `destroy` manually is not typically needed or recommended,
+/// and instead you should use [`define_bytebuffer_destructor!`].
+///
+/// ## Layout/fields
+///
+/// This struct's field are not `pub` (mostly so that we can soundly implement `Send`, but also so
+/// that we can verify Rust users are constructing them appropriately), the fields, their types, and
+/// their order are *very much* a part of the public API of this type. Consumers on the other side
+/// of the FFI will need to know its layout.
+///
+/// If this were a C struct, it would look like
+///
+/// ```c,no_run
+/// struct ByteBuffer {
+///     int64_t len;
+///     uint8_t *data; // note: nullable
+/// };
+/// ```
+///
+/// In Rust, there are two fields, in this order: `len: i32`, and `data: *mut u8`.
+///
+/// ### Description of fields
+///
+/// `data` is a pointer to an array of `len` bytes. Not that data can be a null pointer and therefore
+/// should be checked.
+///
+/// The bytes array is allocated on the heap and must be freed on it as well. Critically, if there
+/// are multiple rust packages using being used in the same application, it *must be freed on the
+/// same heap that allocated it*, or you will corrupt both heaps.
+///
+/// Typically, this object is managed on the other side of the FFI (on the "FFI consumer"), which
+/// means you must expose a function to release the resources of `data` which can be done easily
+/// using the [`define_bytebuffer_destructor!`] macro provided by this crate.
+#[repr(C)]
+pub struct ByteBuffer {
```

Comment:
Would it make sense to expose a i32 and a i64 version upstream instead?

Reply:
I can talk to Thom and see what he thinks about it.
It's simple enough code that I'm not worried about the duplication we have here.
Changing it might be problematic for a-s though.

Reply:
Oh, I see. No problem then!